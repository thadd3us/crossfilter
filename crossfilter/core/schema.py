"""Schema definitions, constants, and enums for crossfilter data structures."""

import json
import logging
import sqlite3
from enum import StrEnum
from pathlib import Path
from typing import Optional

import pandas as pd
import pandera.pandas as pa
from pandera.typing import Series

# Import shared types from backend_frontend_shared_schema
from crossfilter.core.backend_frontend_shared_schema import (
    FilterEvent,
    ProjectionType,
    TemporalLevel,
)

logger = logging.getLogger(__name__)


class SchemaColumns(StrEnum):
    """Column names from the DataSchema."""

    # Used to refer to the DataFrame integer index.
    DF_ID = "df_id"

    UUID_STRING = "UUID_STRING"
    DATA_TYPE = "DATA_TYPE"
    NAME = "NAME"
    CAPTION = "CAPTION"
    SOURCE_FILE = "SOURCE_FILE"
    TIMESTAMP_MAYBE_TIMEZONE_AWARE = "TIMESTAMP_MAYBE_TIMEZONE_AWARE"
    TIMESTAMP_UTC = "TIMESTAMP_UTC"
    GPS_LATITUDE = "GPS_LATITUDE"
    GPS_LONGITUDE = "GPS_LONGITUDE"
    RATING_0_TO_5 = "RATING_0_TO_5"
    SIZE_IN_BYTES = "SIZE_IN_BYTES"
    COUNT = "COUNT"

    # CLIP embedding UMAP projection coordinates (on a sphere like lat/lon)
    CLIP_UMAP_HAVERSINE_LATITUDE = "CLIP_UMAP_HAVERSINE_LATITUDE"
    CLIP_UMAP_HAVERSINE_LONGITUDE = "CLIP_UMAP_HAVERSINE_LONGITUDE"
    
    # SigLIP2 embedding UMAP projection coordinates (on a sphere like lat/lon)
    SIGLIP2_UMAP2D_HAVERSINE_LATITUDE = "SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"
    SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE = "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"


class DataType(StrEnum):
    PHOTO = "PHOTO"
    VIDEO = "VIDEO"
    GPX_TRACKPOINT = "GPX_TRACKPOINT"
    GPX_WAYPOINT = "GPX_WAYPOINT"


class EmbeddingType(StrEnum):
    SIGLIP2 = "SIGLIP2"

class DataSchema(pa.DataFrameModel):
    UUID_STRING: Series[str] = pa.Field(nullable=True)
    DATA_TYPE: Series[str] = pa.Field(
        nullable=True, isin=list(DataType.__members__.values())
    )
    NAME: Series[str] = pa.Field(nullable=True)
    CAPTION: Series[str] = pa.Field(nullable=True)
    SOURCE_FILE: Series[str] = pa.Field(nullable=True)
    TIMESTAMP_MAYBE_TIMEZONE_AWARE: Series[str] = pa.Field(nullable=True)
    TIMESTAMP_UTC: Series[pd.DatetimeTZDtype] = pa.Field(
        nullable=True, dtype_kwargs={"tz": "UTC"}
    )
    GPS_LATITUDE: Series[float] = pa.Field(nullable=True, ge=-90, le=90)
    GPS_LONGITUDE: Series[float] = pa.Field(nullable=True, ge=-180, le=180)
    RATING_0_TO_5: Series[pd.Int64Dtype] = pa.Field(nullable=True, ge=0, le=5)
    SIZE_IN_BYTES: Series[pd.Int64Dtype] = pa.Field(nullable=True, ge=0)

    # For aggregated data, the COUNT column says how many rows are aggregated into the bucket.
    # Some of the other columns may still be present, coming from a single row sample of the data aggregated into the bucket.
    # For single point data, COUNT is missing and implied to be 1.
    COUNT: Optional[Series[pd.Int64Dtype]]  # = pa.Field(nullable=True, ge=0)

    class Config:
        strict = True
        coerce = True


def get_h3_column_name(level: int) -> str:
    """Get the H3 column name for a specific level (0-15)."""
    if not 0 <= level <= 15:
        raise ValueError(f"H3 level must be between 0 and 15, got {level}")
    return f"BUCKETED_H3_L{level}"


def get_clip_umap_h3_column_name(level: int) -> str:
    """Get the CLIP UMAP H3 column name for a specific level (0-15)."""
    if not 0 <= level <= 15:
        raise ValueError(f"H3 level must be between 0 and 15, got {level}")
    return f"BUCKETED_CLIP_HAVERSINE_UMAP_H3_L{level}"


def get_temporal_column_name(level: TemporalLevel) -> str:
    """Get the temporal column name for a specific temporal level."""
    return f"BUCKETED_TIMESTAMP_{level}"


required_columns = [
    SchemaColumns.UUID_STRING,
    SchemaColumns.DATA_TYPE,
    SchemaColumns.NAME,
    SchemaColumns.CAPTION,
    SchemaColumns.SOURCE_FILE,
    SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
    SchemaColumns.TIMESTAMP_UTC,
    SchemaColumns.GPS_LATITUDE,
    SchemaColumns.GPS_LONGITUDE,
    SchemaColumns.RATING_0_TO_5,
    SchemaColumns.SIZE_IN_BYTES,
]


def load_jsonl_to_dataframe(jsonl_path: Path) -> pd.DataFrame:
    """
    Load a JSONL file into a DataFrame conforming to DataSchema.

    The JSONL file may have extra columns and may not specify all required columns.
    Missing schema columns will be created with null values, and data will be coerced to
    the correct types according to the schema. Extra columns are preserved in the output.

    Args:
        jsonl_path: Path to the JSONL file

    Returns:
        DataFrame with schema columns validated, plus any extra columns from JSONL
    """

    if not jsonl_path.exists():
        raise FileNotFoundError(f"JSONL file not found: {jsonl_path}")

    # Read JSONL file
    records = []
    with open(jsonl_path, encoding="utf-8") as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            try:
                record = json.loads(line)
                records.append(record)
            except json.JSONDecodeError as e:
                raise ValueError(f"Invalid JSON on line {line_num}: {e}")

    df = pd.DataFrame(records)

    # Ensure all required columns exist, adding missing ones with null values

    for col in required_columns:
        if col not in df.columns:
            df[col] = None

    # Convert SOURCE_FILE from Path objects to strings, handling missing values
    df[SchemaColumns.SOURCE_FILE] = df[SchemaColumns.SOURCE_FILE].astype(str)

    # Convert TIMESTAMP_UTC to UTC timezone-aware datetime, handling missing values
    df[SchemaColumns.TIMESTAMP_UTC] = pd.to_datetime(
        df[SchemaColumns.TIMESTAMP_UTC], utc=True, errors="coerce"
    )

    # Validate and coerce only the schema columns
    schema_df = df[required_columns].copy()
    validated_schema_df = DataSchema.validate(schema_df, lazy=True)

    # Combine validated schema columns with any extra columns from original data
    extra_columns = [col for col in df.columns if col not in required_columns]
    if extra_columns:
        result_df = pd.concat([validated_schema_df, df[extra_columns]], axis=1)
    else:
        result_df = validated_schema_df

    # Set stable df_id index using pandas int64 index
    result_df = result_df.reset_index(drop=True)
    result_df.index.name = SchemaColumns.DF_ID

    logger.info(f"Loaded {len(result_df)} records from {jsonl_path=}")

    return result_df


def load_sqlite_to_dataframe(sqlite_db_path: Path, table_name: str) -> pd.DataFrame:
    if not sqlite_db_path.exists():
        raise FileNotFoundError(f"SQLite database not found: {sqlite_db_path}")

    # Read data from SQLite table
    with sqlite3.connect(sqlite_db_path) as conn:
        df = pd.read_sql(f"SELECT * FROM {table_name}", conn)

    for col in required_columns:
        if col not in df.columns:
            df[col] = None

    # Convert TIMESTAMP_UTC to UTC timezone-aware datetime, handling missing values
    # THAD: FIXME -- why is this needed?
    df[SchemaColumns.TIMESTAMP_UTC] = pd.to_datetime(
        df[SchemaColumns.TIMESTAMP_UTC], utc=True, errors="coerce"
    )
    df[SchemaColumns.GPS_LATITUDE] = df[SchemaColumns.GPS_LATITUDE].clip(-90, 90)
    df[SchemaColumns.GPS_LONGITUDE] = df[SchemaColumns.GPS_LONGITUDE].clip(-180, 180)

    df[SchemaColumns.CLIP_UMAP_HAVERSINE_LATITUDE] = df[
        SchemaColumns.CLIP_UMAP_HAVERSINE_LATITUDE
    ].clip(-90, 90)
    df[SchemaColumns.CLIP_UMAP_HAVERSINE_LONGITUDE] = df[
        SchemaColumns.CLIP_UMAP_HAVERSINE_LONGITUDE
    ].clip(-180, 180)

    # Validate and coerce only the schema columns
    schema_df = df[required_columns].copy()
    # THAD: FIXME.
    validated_schema_df = schema_df.copy()  # DataSchema.validate(schema_df, lazy=True)

    # Combine validated schema columns with any extra columns from original data
    extra_columns = [col for col in df.columns if col not in required_columns]
    if extra_columns:
        result_df = pd.concat([validated_schema_df, df[extra_columns]], axis=1)
    else:
        result_df = validated_schema_df

    logger.info(
        f"Loaded {len(result_df)} records from {sqlite_db_path=} table '{table_name}'"
    )

    return result_df
