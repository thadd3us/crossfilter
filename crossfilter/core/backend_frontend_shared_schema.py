"""Shared schema definitions for backend-frontend communication."""

from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Optional

from pydantic import BaseModel, Field


class ProjectionType(StrEnum):
    """Types of data projections supported by crossfilter."""
    
    TEMPORAL = "temporal"
    GEO = "geo"
    CLIP_EMBEDDING = "clip_embedding"  # placeholder for future


class FilterOperatorType(StrEnum):
    """Types of filter operations supported by crossfilter."""
    
    INTERSECTION = "intersection"  # Limit to current selection
    SUBTRACTION = "subtraction"    # Remove current selection


@dataclass(frozen=True)
class FilterEvent:
    """Represents a filter event from a specific projection."""
    
    projection_type: ProjectionType
    selected_df_ids: set[int]
    filter_operator: FilterOperatorType


# API Request Models
class LoadDataRequest(BaseModel):
    """Request model for loading data."""
    
    file_path: str


class DfIdsFilterRequest(BaseModel):
    """Request model for filtering to specific df_ids from a plot."""
    
    df_ids: list[int] = Field(..., description="df_ids from the plot (could be from lasso selection, visible area, etc.)")
    event_source: ProjectionType = Field(..., description="Which plot type this filtering comes from")
    filter_operator: FilterOperatorType = Field(..., description="Type of filter operation to apply")


# API Response Models
class SessionStateResponse(BaseModel):
    """Response model for session state information."""
    
    all_rows_count: int
    filtered_rows_count: int
    columns: list[str]
    memory_usage_mb: str
    temporal_projection: dict[str, Any]
    geo_projection: dict[str, Any]
    # Frontend-compatible aliases
    has_data: bool
    row_count: int
    filtered_count: int


class LoadDataResponse(BaseModel):
    """Response model for data loading operations."""
    
    success: bool
    message: str
    session_state: SessionStateResponse


class FilterResponse(BaseModel):
    """Response model for filter operations."""
    
    success: bool
    filter_state: SessionStateResponse


class TemporalPlotResponse(BaseModel):
    """Response model for temporal plot data."""
    
    plotly_plot: dict[str, Any]
    data_type: str
    point_count: int
    distinct_point_count: int
    aggregation_level: Optional[str]


class GeoPlotResponse(BaseModel):
    """Response model for geo plot data."""
    
    plotly_plot: dict[str, Any]
    marker_count: int
    distinct_point_count: int
    aggregation_level: Optional[str]


# Server-Sent Events Models
class SSEEvent(BaseModel):
    """Base model for Server-Sent Events."""
    
    type: str
    timestamp: float
    version: int


class FilterChangeEvent(SSEEvent):
    """Server-Sent Event for filter changes."""
    
    affected_components: list[str]
    session_state: SessionStateResponse


class HeartbeatEvent(SSEEvent):
    """Server-Sent Event for heartbeat."""
    
    pass


class ConnectionEstablishedEvent(FilterChangeEvent):
    """Server-Sent Event for connection establishment."""
    
    pass


