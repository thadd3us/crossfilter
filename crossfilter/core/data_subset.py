from dataclasses import dataclass

import pandas as pd


@dataclass
class DataSubset:
    subset_uuid: str

    df: pd.DataFrame
