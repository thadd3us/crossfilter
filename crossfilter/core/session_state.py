"""Session state management for single-session Crossfilter application."""

import logging

import pandas as pd

from crossfilter.core.bucketing import (
    H3_LEVELS,
    add_bucketed_columns,
    bucket_by_target_column,
    get_optimal_h3_level,
    get_optimal_temporal_level,
)
from crossfilter.core.filter_state import FilterState
from crossfilter.core.schema import (
    SchemaColumns,
    TemporalLevel,
    get_h3_column_name,
    get_temporal_column_name,
)

logger = logging.getLogger(__name__)


class SessionState:
    """
    Manages the current state of data loaded into the Crossfilter application.

    This class represents the single session state for the application, holding
    the currently loaded dataset and any derived data structures needed for
    crossfiltering and visualization.

    In the single-session design pattern, there is exactly one instance of this
    class per web server instance, simplifying state management and eliminating
    the need for complex multi-user session handling.
    """

    def __init__(self) -> None:
        """Initialize session state with empty DataFrame."""
        self._data = pd.DataFrame()
        self._bucketed_data = pd.DataFrame()
        self._filter_state = FilterState()
        self._update_metadata()

    @property
    def data(self) -> pd.DataFrame:
        """Get the current dataset."""
        return self._data

    @data.setter
    def data(self, value: pd.DataFrame) -> None:
        """Set the current dataset."""
        self._data = value
        self._bucketed_data = add_bucketed_columns(value)
        self._filter_state.initialize_with_data(self._bucketed_data)
        self._update_metadata()
        logger.info(f"Loaded dataset with {len(value)} rows into session state")

    def _update_metadata(self) -> None:
        """Update metadata based on current DataFrame."""
        self._metadata = {
            "shape": self._data.shape,
            "columns": list(self._data.columns),
            "dtypes": self._data.dtypes.to_dict(),
        }

    @property
    def metadata(self) -> dict:
        """Get metadata about the current dataset."""
        return self._metadata.copy()

    def has_data(self) -> bool:
        """Check if the session has data loaded."""
        return not self._data.empty

    def clear(self) -> None:
        """Clear all data from the session state."""
        self._data = pd.DataFrame()
        self._bucketed_data = pd.DataFrame()
        self._filter_state = FilterState()
        self._update_metadata()

    def load_dataframe(self, df: pd.DataFrame) -> None:
        """Load a DataFrame into the session state."""
        self.data = df

    def get_summary(self) -> dict:
        """Get a summary of the current session state."""
        if not self.has_data():
            return {"status": "empty", "message": "No data loaded"}

        summary = {
            "status": "loaded",
            "shape": self._metadata["shape"],
            "columns": self._metadata["columns"],
            "memory_usage": f"{self._data.memory_usage(deep=True).sum() / 1024 / 1024:.2f} MB",
        }

        # Add filter state info
        summary["filter_state"] = self._filter_state.get_summary()

        return summary

    @property
    def bucketed_data(self) -> pd.DataFrame:
        """Get the bucketed dataset."""
        return self._bucketed_data

    @property
    def filter_state(self) -> FilterState:
        """Get the filter state manager."""
        return self._filter_state

    def get_filtered_data(self) -> pd.DataFrame:
        """Get the currently filtered dataset."""
        return self._filter_state.get_filtered_dataframe(self._bucketed_data)

    def get_spatial_aggregation(self, max_groups: int) -> pd.DataFrame:
        """
        Get spatially aggregated data for visualization.

        Args:
            max_groups: Maximum number of groups to return

        Returns:
            Aggregated DataFrame suitable for heatmap visualization
        """
        filtered_data = self.get_filtered_data()

        if len(filtered_data) <= max_groups:
            # Return individual points if under threshold
            columns = [SchemaColumns.GPS_LATITUDE, SchemaColumns.GPS_LONGITUDE]
            result = filtered_data[columns].copy()
            # Add df_id column for consistency
            result[SchemaColumns.DF_ID] = filtered_data.index
            return result

        # Find optimal H3 level
        optimal_level = get_optimal_h3_level(filtered_data, max_groups)

        if optimal_level is None:
            # Fallback to least granular level
            optimal_level = min(H3_LEVELS)

        # Aggregate by H3 cells
        h3_column = get_h3_column_name(optimal_level)
        bucketed = bucket_by_target_column(filtered_data, h3_column)

        # Rename COUNT column to count for consistency
        result = bucketed.rename(columns={"COUNT": "count"})
        return result

    def get_temporal_aggregation(self, max_groups: int) -> pd.DataFrame:
        """
        Get temporally aggregated data for CDF visualization.

        Args:
            max_groups: Maximum number of groups to return

        Returns:
            Aggregated DataFrame suitable for CDF visualization
        """
        filtered_data = self.get_filtered_data()

        if len(filtered_data) <= max_groups:
            # Return individual points if under threshold
            df = filtered_data[[SchemaColumns.TIMESTAMP_UTC]].copy()
            df = df.sort_values(SchemaColumns.TIMESTAMP_UTC)
            df["cumulative_count"] = range(1, len(df) + 1)
            # Add df_id column for consistency
            df[SchemaColumns.DF_ID] = df.index
            return df

        # Find optimal temporal level
        optimal_level = get_optimal_temporal_level(filtered_data, max_groups)

        if optimal_level is None:
            # Fallback to least granular level
            optimal_level = TemporalLevel.YEAR

        # Aggregate by temporal buckets
        temporal_column = get_temporal_column_name(optimal_level)
        bucketed = bucket_by_target_column(filtered_data, temporal_column)

        # Transform to match expected output format for CDF visualization
        result = bucketed.rename(columns={"COUNT": "count"})

        # Sort by timestamp and add cumulative count for CDF
        result = result.sort_values(temporal_column)
        result["cumulative_count"] = result["count"].cumsum()

        return result
