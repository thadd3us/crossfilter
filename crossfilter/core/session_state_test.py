"""Tests for session state management."""

import pandas as pd
import pytest

from crossfilter.core.schema import FilterEvent, ProjectionType, SchemaColumns as C
from crossfilter.core.session_state import SessionState


@pytest.fixture
def sample_df() -> pd.DataFrame:
    """Create a sample DataFrame for testing."""
    df = pd.DataFrame(
        {
            C.UUID_STRING: [f"uuid_{i}" for i in range(20)],
            C.GPS_LATITUDE: [37.7749 + i * 0.001 for i in range(20)],
            C.GPS_LONGITUDE: [-122.4194 + i * 0.001 for i in range(20)],
            C.TIMESTAMP_UTC: [f"2024-01-01T{10 + i // 4}:00:00Z" for i in range(20)],
        }
    )
    # Convert timestamp to datetime
    df[C.TIMESTAMP_UTC] = pd.to_datetime(df[C.TIMESTAMP_UTC], utc=True)
    # Set stable df_id index
    df.index.name = C.DF_ID
    return df


def test_session_state_initialization() -> None:
    """Test SessionState initialization."""
    session = SessionState()

    assert not session.has_data()
    assert session.all_rows.empty
    assert session.filtered_rows.empty

    summary = session.get_summary()
    assert summary["status"] == "empty"
    assert summary["message"] == "No data loaded"


def test_load_dataframe(sample_df: pd.DataFrame) -> None:
    """Test loading a DataFrame into session state."""
    session = SessionState()
    session.load_dataframe(sample_df)

    assert session.has_data()
    assert len(session.all_rows) == 20
    assert len(session.filtered_rows) == 20
    assert session.all_rows.index.name == C.DF_ID
    assert session.filtered_rows.index.name == C.DF_ID

    # Check that bucketed columns were added
    all_rows = session.all_rows
    assert any(col.startswith("QUANTIZED_H3_L") for col in all_rows.columns)
    assert any(col.startswith("QUANTIZED_TIMESTAMP_") for col in all_rows.columns)


def test_session_state_metadata(sample_df: pd.DataFrame) -> None:
    """Test metadata tracking."""
    session = SessionState()
    session.load_dataframe(sample_df)

    metadata = session.metadata
    # Shape includes bucketed columns now
    assert metadata["shape"][0] == 20  # Same number of rows
    assert metadata["shape"][1] > 4   # More columns due to bucketing
    assert C.GPS_LATITUDE in metadata["columns"]
    assert C.TIMESTAMP_UTC in metadata["columns"]

    summary = session.get_summary()
    assert summary["status"] == "loaded"
    assert summary["all_rows_count"] == 20
    assert summary["filtered_rows_count"] == 20
    assert "memory_usage" in summary
    assert "temporal_projection" in summary
    assert "geo_projection" in summary


def test_projection_states(sample_df: pd.DataFrame) -> None:
    """Test that projection states are properly initialized and updated."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Check temporal projection
    temporal_proj = session.temporal_projection
    assert temporal_proj.max_rows == 100000  # default
    assert len(temporal_proj.projection_df) == 20
    assert temporal_proj.current_aggregation_level is None  # individual points

    # Check geo projection
    geo_proj = session.geo_projection
    assert geo_proj.max_rows == 100000  # default
    assert len(geo_proj.projection_df) == 20
    assert geo_proj.current_h3_level is None  # individual points


def test_spatial_aggregation_individual_points(sample_df: pd.DataFrame) -> None:
    """Test spatial aggregation when under threshold."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Request more groups than we have points
    result = session.get_spatial_aggregation(max_groups=100)

    # Should return individual points
    assert len(result) == 20
    assert C.GPS_LATITUDE in result.columns
    assert C.GPS_LONGITUDE in result.columns
    assert C.DF_ID in result.columns
    # Should not have COUNT column for individual points
    assert "count" not in result.columns


def test_spatial_aggregation_aggregated(sample_df: pd.DataFrame) -> None:
    """Test spatial aggregation when over threshold."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Request fewer groups than we have points
    result = session.get_spatial_aggregation(max_groups=5)

    # Should return aggregated data
    assert len(result) <= 5
    assert C.GPS_LATITUDE in result.columns
    assert C.GPS_LONGITUDE in result.columns
    assert "count" in result.columns

    # Total count should match original data
    assert result["count"].sum() == 20


def test_temporal_aggregation_individual_points(sample_df: pd.DataFrame) -> None:
    """Test temporal aggregation when under threshold."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Request more groups than we have points
    result = session.get_temporal_aggregation(max_groups=100)

    # Should return individual points
    assert len(result) == 20
    assert C.TIMESTAMP_UTC in result.columns
    assert "cumulative_count" in result.columns
    assert C.DF_ID in result.columns
    # Should not have COUNT column for individual points
    assert "count" not in result.columns

    # Should be sorted by timestamp
    assert result[C.TIMESTAMP_UTC].is_monotonic_increasing

    # Cumulative count should go from 1 to 20
    assert list(result["cumulative_count"]) == list(range(1, 21))


def test_temporal_aggregation_aggregated(sample_df: pd.DataFrame) -> None:
    """Test temporal aggregation when over threshold."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Request fewer groups than we have points
    result = session.get_temporal_aggregation(max_groups=3)

    # Should return aggregated data
    assert len(result) <= 3
    assert "count" in result.columns
    assert "cumulative_count" in result.columns

    # Total count should match original data
    assert result["count"].sum() == 20

    # Cumulative count should end at total
    assert result["cumulative_count"].iloc[-1] == 20


def test_apply_filter_event_temporal(sample_df: pd.DataFrame) -> None:
    """Test applying a temporal filter event."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Get initial counts
    assert len(session.filtered_rows) == 20
    
    # Apply a temporal filter (select first 10 points)
    selected_df_ids = set(range(0, 10))
    filter_event = FilterEvent(ProjectionType.TEMPORAL, selected_df_ids)
    session.apply_filter_event(filter_event)

    # Check that filtered_rows was updated
    assert len(session.filtered_rows) == 10
    assert all(idx in range(0, 10) for idx in session.filtered_rows.index)

    # Check that projections were updated
    temporal_result = session.get_temporal_aggregation(max_groups=100)
    assert len(temporal_result) == 10


def test_apply_filter_event_geo(sample_df: pd.DataFrame) -> None:
    """Test applying a geographic filter event."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Get initial counts
    assert len(session.filtered_rows) == 20
    
    # Apply a geographic filter (select first 10 points)
    selected_df_ids = set(range(0, 10))
    filter_event = FilterEvent(ProjectionType.GEO, selected_df_ids)
    session.apply_filter_event(filter_event)

    # Check that filtered_rows was updated
    assert len(session.filtered_rows) == 10
    assert all(idx in range(0, 10) for idx in session.filtered_rows.index)

    # Check that projections were updated
    geo_result = session.get_spatial_aggregation(max_groups=100)
    assert len(geo_result) == 10


def test_reset_filters(sample_df: pd.DataFrame) -> None:
    """Test resetting filters."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Apply a filter first
    selected_df_ids = set(range(0, 10))
    filter_event = FilterEvent(ProjectionType.TEMPORAL, selected_df_ids)
    session.apply_filter_event(filter_event)
    assert len(session.filtered_rows) == 10

    # Reset filters
    session.reset_filters()

    # Should be back to all data
    assert len(session.filtered_rows) == 20
    assert len(session.filtered_rows) == len(session.all_rows)


def test_clear_session_state(sample_df: pd.DataFrame) -> None:
    """Test clearing session state."""
    session = SessionState()
    session.load_dataframe(sample_df)

    assert session.has_data()

    session.clear()

    assert not session.has_data()
    assert session.all_rows.empty
    assert session.filtered_rows.empty

    summary = session.get_summary()
    assert summary["status"] == "empty"


def test_empty_dataframe_handling() -> None:
    """Test handling of empty DataFrames."""
    session = SessionState()
    empty_df = pd.DataFrame(
        columns=[
            C.GPS_LATITUDE,
            C.GPS_LONGITUDE,
            C.TIMESTAMP_UTC,
        ]
    )
    empty_df.index.name = C.DF_ID

    session.load_dataframe(empty_df)

    assert not session.has_data()  # Empty DataFrame

    # Aggregation should handle empty data gracefully
    spatial_result = session.get_spatial_aggregation(max_groups=100)
    temporal_result = session.get_temporal_aggregation(max_groups=100)

    assert len(spatial_result) == 0
    assert len(temporal_result) == 0


def test_filter_integration_with_aggregation(sample_df: pd.DataFrame) -> None:
    """Test that filtering affects aggregation results."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Get initial aggregation
    initial_spatial = session.get_spatial_aggregation(max_groups=100)
    initial_temporal = session.get_temporal_aggregation(max_groups=100)

    assert len(initial_spatial) == 20
    assert len(initial_temporal) == 20

    # Apply filter to reduce to 10 points
    filtered_df_ids = set(range(0, 10))
    filter_event = FilterEvent(ProjectionType.TEMPORAL, filtered_df_ids)
    session.apply_filter_event(filter_event)

    # Get aggregation after filtering
    filtered_spatial = session.get_spatial_aggregation(max_groups=100)
    filtered_temporal = session.get_temporal_aggregation(max_groups=100)

    assert len(filtered_spatial) == 10
    assert len(filtered_temporal) == 10

    # All df_ids should be in the filtered set
    assert all(df_id in filtered_df_ids for df_id in filtered_spatial[C.DF_ID])
    assert all(df_id in filtered_df_ids for df_id in filtered_temporal[C.DF_ID])


def test_projection_max_rows_updates(sample_df: pd.DataFrame) -> None:
    """Test that max_rows parameter updates projections correctly."""
    session = SessionState()
    session.load_dataframe(sample_df)

    # Initially should show individual points
    temporal_result = session.get_temporal_aggregation(max_groups=100)
    assert "count" not in temporal_result.columns  # Individual points

    # Change to force aggregation
    temporal_result = session.get_temporal_aggregation(max_groups=5)
    assert "count" in temporal_result.columns  # Aggregated

    # Verify projection state was updated
    assert session.temporal_projection.max_rows == 5


def test_unknown_projection_type(sample_df: pd.DataFrame) -> None:
    """Test handling of unknown projection types."""
    session = SessionState()
    session.load_dataframe(sample_df)

    initial_count = len(session.filtered_rows)
    
    # Try to apply filter with unknown projection type
    filter_event = FilterEvent(ProjectionType.CLIP_EMBEDDING, set(range(0, 10)))
    session.apply_filter_event(filter_event)
    
    # Should not have changed the filtered data
    assert len(session.filtered_rows) == initial_count