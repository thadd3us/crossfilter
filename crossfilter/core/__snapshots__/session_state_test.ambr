# serializer version: 1
# name: test_temporal_aggregation_aggregated
  list([
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 10,
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.GPX_WAYPOINT: 'GPX_WAYPOINT'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7749,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.4194,
      'QUANTIZED_H3_L0': '8029fffffffffff',
      'QUANTIZED_H3_L1': '81283ffffffffff',
      'QUANTIZED_H3_L10': '8a283082800ffff',
      'QUANTIZED_H3_L11': '8b283082800bfff',
      'QUANTIZED_H3_L12': '8c283082800b3ff',
      'QUANTIZED_H3_L13': '8d283082800b3bf',
      'QUANTIZED_H3_L14': '8e283082800b397',
      'QUANTIZED_H3_L15': '8f283082800b390',
      'QUANTIZED_H3_L2': '822837fffffffff',
      'QUANTIZED_H3_L3': '832830fffffffff',
      'QUANTIZED_H3_L4': '8428309ffffffff',
      'QUANTIZED_H3_L5': '85283083fffffff',
      'QUANTIZED_H3_L6': '86283082fffffff',
      'QUANTIZED_H3_L7': '872830828ffffff',
      'QUANTIZED_H3_L8': '8828308281fffff',
      'QUANTIZED_H3_L9': '89283082803ffff',
      'QUANTIZED_TIMESTAMP_DAY': Timestamp('2024-01-01 00:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_HOUR': Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_MINUTE': Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_MONTH': Timestamp('2024-01-01 00:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_SECOND': Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_YEAR': Timestamp('2024-01-01 00:00:00+0000', tz='UTC'),
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: 'uuid_0',
    }),
    dict({
      <SchemaColumns.COUNT: 'COUNT'>: 10,
      <SchemaColumns.DATA_TYPE: 'DATA_TYPE'>: <DataType.PHOTO: 'PHOTO'>,
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>: 37.7759,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>: -122.41839999999999,
      'QUANTIZED_H3_L0': '8029fffffffffff',
      'QUANTIZED_H3_L1': '81283ffffffffff',
      'QUANTIZED_H3_L10': '8a28308280e7fff',
      'QUANTIZED_H3_L11': '8b28308280e2fff',
      'QUANTIZED_H3_L12': '8c28308280e27ff',
      'QUANTIZED_H3_L13': '8d28308280e277f',
      'QUANTIZED_H3_L14': '8e28308280e2777',
      'QUANTIZED_H3_L15': '8f28308280e2775',
      'QUANTIZED_H3_L2': '822837fffffffff',
      'QUANTIZED_H3_L3': '832830fffffffff',
      'QUANTIZED_H3_L4': '8428309ffffffff',
      'QUANTIZED_H3_L5': '85283083fffffff',
      'QUANTIZED_H3_L6': '86283082fffffff',
      'QUANTIZED_H3_L7': '872830828ffffff',
      'QUANTIZED_H3_L8': '8828308281fffff',
      'QUANTIZED_H3_L9': '8928308280fffff',
      'QUANTIZED_TIMESTAMP_DAY': Timestamp('2024-01-01 00:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_HOUR': Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_MINUTE': Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_MONTH': Timestamp('2024-01-01 00:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_SECOND': Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      'QUANTIZED_TIMESTAMP_YEAR': Timestamp('2024-01-01 00:00:00+0000', tz='UTC'),
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: Timestamp('2024-01-01 10:00:00+0000', tz='UTC'),
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: 'uuid_1',
    }),
  ])
# ---
